### PR TITLE
fix(relayer): end a claim's send at its nonce's holder, and ack duplicate deliveries

### PR DESCRIPTION
**Pull request by davidtaikocha**
Two defects that #22070 could only expose once it was carrying real claims. `DEST_PRIVATE_RPC_URLS` was enabled on the mainnet L2→L1 processor on 2026-09-01 at 01:57:05Z — a config-only change on an image that already had the code — and the first claim through it found both.

The claim itself succeeded, and cleanly: message `0xb1395af059080682051ce303d3f63804d6d73d26fd31bf6f6c3290e1c0b2a5e6` was accepted by Flashbots, mined in L1 block 25879543 as `0x7b4218cee205c5089cba4a39f26ed4a800a9d8cb4ec786203a23a023c4693bb1`, and `messageStatus` is `2`. Final counters were `sends=3/0 fail=0/1 held=1/0 trips=0/0 unavailable=0 all_refused=0 processed=1 reverted=0` — the claim never reached the public mempool, which is the guarantee the feature exists for. The `fail=0/1` is the first defect below.

## A benign tail race walks a healthy backup endpoint out of rotation

```
02:48:23  tx 0x7b4218ce… mined in L1 block 25879543
02:48:35  Private endpoint already holds this nonce  endpoint=0 nonce=7210
02:48:36  Private endpoint refused a transaction     endpoint=1 txHash=0x7b4218ce…
          error="Failed in pending block with: Reverted"
02:48:36  Mined tx  txHash=7b4218ce…
```

The transaction manager's next resubmission was already in flight when the receipt landed. It went to endpoint 0 first, correctly, because endpoint 0 holds nonce 7210, and Flashbots answered that it still holds it. The send then **continued to endpoint 1**, which simulated the transaction against the pending block, saw the now-DONE message revert, and refused. MEV Blocker did nothing wrong and was charged a failure anyway.

That charge sticks. In `recordFailure`, `consecutive[index]++` runs before the same-nonce deduplication, and the deduplication would not apply regardless because every claim carries a distinct nonce. Both counters are cleared only by `recordSuccess`, which fires only when *that* endpoint accepts something — and in steady state endpoint 0 always accepts first, so endpoint 1 never gets a reset. At `DefaultPrivateRPCFailureThreshold = 3`, **three claims take MEV Blocker out of rotation for five minutes**. It returns, and three more take it out again. At the observed L2→L1 volume of roughly seven claims a day, about twice a day.

Four consequences, in rough order of how much they matter:

1. The backup spends much of its life out of rotation, so when the primary genuinely fails there may be no private endpoint left and claims go out through `DEST_RPC_URL` — the exposure this whole feature removes.
2. `private_rpc_trips`, documented as "the transition worth alerting on", fires on non-outages.
3. The same signed transaction is left live in two builders' pools, which #22070's own description argues against: "sending it elsewhere leaves both live in different builders' pools."
4. `SendTransaction` returns endpoint 1's rejection, so the transaction manager is told the publish failed while endpoint 0 is demonstrably carrying the claim. A non-nil return does not arm `bumpFees` and does not count toward `successFullPublishCount`, which gates `TxNotInMempoolTimeout`.

### The fix

`recordHeldNonce` now reports, as a second return value, whether the answering endpoint actually took this nonce. The send ends there rather than offering the claim to the endpoints behind it. The predicate stays inside `recordHeldNonce` rather than becoming a second query function: it is the one place that owns this state, and a copy would drift.

That question is answered from what the endpoint actually accepted, not from the `highestAccepted` high-water mark. The mark is right for the two jobs it already had — exempting the accounting, and ordering the holder first — and wrong for this one. Claims are signed in nonce order but sent concurrently, so a later nonce can be accepted while an earlier one never was; the mark then covers a nonce nobody here has seen. Being generous costs nothing when the consequence is "don't charge this endpoint", and strands the claim with nobody holding it until `TX_SEND_TIMEOUT` when the consequence is "stop the send". So the exemption and the ordering keep the mark, and the terminal decision reads an exact record — bounded at 256 and evicting the stalest, mirroring the refusal counts, because a claim abandoned at `TX_SEND_TIMEOUT` never comes back to clear its entry.

That record stores `claimIdentity` alongside each nonce, and both have to match, for the same reason `refusalRun` carries a claim hash. Nonces recycle: the abandoned claim leaves its entry behind and `resetNonce` hands the nonce to an unrelated message, which would otherwise inherit the acceptance — the endpoint answering `replacement transaction underpriced` about the claim it is still holding would end the new claim's send with nobody carrying it. The identity is the calldata, stable across the fee bumps within one send and different for every other claim, so the holder path itself is unaffected.

The acceptance is read *before* the generation guard, so it survives one. A concurrent send can trip an endpoint while our held answer is still on the wire; that says the endpoint is unhealthy, not that it stopped carrying what it took. The guard exists to stop a stale result from spending the fresh budget — it suppresses the accounting and leaves the fact alone.

For a nonce the endpoint never accepted — the gossip case, where it may have learned the claim from another relay — the existing `continue` is preserved, because the endpoints behind it genuinely have not been asked.

Deliberately unchanged: `recordSuccess` is not called (it clears `consecutive` and `failures`, defeating the ceiling `recordHeldNonce` exists to enforce); `private_rpc_sends` is not incremented (no new transaction was accepted, and `private_rpc_held_nonce` already records this path); `clearAllRefused` is not touched (a held answer already sets `allAnsweredRejection = false`, so this path never contributes to a run).

One behaviour genuinely narrows: if a holder accepts a nonce and then goes silent, that claim no longer falls through to a sibling. This is the case `TX_SEND_TIMEOUT` was introduced for — "a relay can accept a transaction that never gets included, and acceptance is the only signal this layer gets" — so the send ends at the five-minute bound and the message is requeued as transient. Held answers for nonces the endpoint never accepted remain bounded by the consecutive ceiling.

## `errAlreadyProcessing` was dead-lettered

The crawler runs with `WATCH_MODE=crawl-past-blocks` and rescans ~50k blocks every ~10 seconds, republishing every message still in `NEW` status. A claim parked in `waitHeaderSynced` — 25m34s for the claim above — is therefore redelivered every ten seconds, each time with a fresh `msgId`:

```
02:37:11 rabbitmq message found msgId=829c946b…
02:37:11 already processing txHash txhash=0x01a64b21…
02:37:11 ERROR process message failed err="already processing txHash"
02:37:11 negatively acknowledged rabbitmq message msgId=829c946b… requeue=false
```

`errAlreadyProcessing` is neither `errUnprocessable` nor transient, so it fell to `default:` and was nacked with `requeue=false`.

It means one thing: this replica is already working on this message. The duplicate carries nothing and the in-flight delivery is authoritative, so it is acked. That is safe because the claim is released on every exit path by a `defer` registered the moment it is taken, and the map is rebuilt at startup — a message cannot become stuck and silently swallow every future delivery.

Acking rather than dead-lettering also makes the guard's key matter, so it moves from the source transaction hash to the message hash. One transaction can emit several `MessageSent` events — a contract bridging more than one token does exactly that — and the indexer publishes one queue message per event, so siblings arrive as separate deliveries sharing a transaction hash. Keyed on that hash the second sibling looked like a duplicate; that was survivable while such a delivery went to a dead-letter queue the broker was dropping anyway, and it stops being survivable once a duplicate is acked. The guard moves out of `processMessage` into `beginProcessing`/`endProcessing` so the key is derived in one place and a test can claim a message the same way production does.

The log drops to debug, because an expected event happening six times a minute is not a warning, and a new `message_sent_events_duplicate_delivery_ops_total` counter takes over the job of making a republish storm visible.

### The same window proved the dead-letter route is dropping messages

Eleven `Nack(requeue=false)` calls between 02:37:00 and 02:39:03 left `dlx-167000-1-MessageSent-queue` pinned at exactly 418 messages. This confirms #22070's account of the routing-key bug on live infrastructure; the 418 are historical residue. The fix shipped there is still inactive on mainnet because the pre-existing queue kept its own arguments — the startup 406 says so and names the remedy.

## Ordering constraint

**This must land before an operator applies the `dead-letter-routing-key` policy.** Today the duplicate rejections above are silently discarded. Once the policy makes dead-lettering work, they would instead accumulate in `dlx-167000-1-MessageSent-queue` at roughly six per minute for every claim parked in `waitHeaderSynced` — about 150 per 25-minute checkpoint wait.

## Out of scope

The crawler's rescan-and-republish cadence is unchanged. Duplicate deliveries still happen; they are now discarded cleanly and counted.

## Testing

Nine new tests in `sending_backend_test.go` and three across `processor_test.go` / `process_message_test.go`, all under `-race`, following each file's existing style.

- A holder answering held-nonce for a nonce it took ends the send: `SendTransaction` returns nil and the backup is never asked. The backup is deliberately healthy, because `fakeSender` records a transaction only when it takes one — a refusing backup leaves `sent` empty whether or not it was asked, which would assert nothing.
- Three claims each ending in a holder's held-nonce answer leave the backup in rotation with `failures` at zero, and the last resend returns nil.
- A held answer for a nonce the endpoint never took still continues to the next endpoint.
- A nonce below the high-water mark that this endpoint never accepted — the out-of-order case — still reaches the backup.
- An endpoint tripped while its held answer was in flight still reports the nonce it accepted, without the stale admission spending anything.
- A nonce recycled to a different claim — same number, different calldata — still reaches the backup when the endpoint answers about the claim it is still holding.
- A hold for someone else's claim under a recycled nonce no longer shields this claim from the public fallback, while a hold for a nonce the endpoint never took still does.
- The accepted-nonce record stays bounded and evicts the stalest rather than the newest.
- A duplicate delivery acks, never nacks, and moves the counter by one.
- A sibling message from the same transaction is not treated as a duplicate, while a redelivery of the same message is.

Each behavioural fix is mutation-checked, and every mutation fails only the test that pins it: reverting the termination to an unconditional `continue` fails the two claim-level tests; making every held answer end the send fails the gossip test along with five existing ones; swapping the exact record back for the high-water mark fails only the out-of-order test; moving the acceptance read back behind the generation guard fails only the tripped-holder test; matching on the nonce without the claim identity fails only the recycled-nonce test; clearing `allAnsweredRejection` unconditionally fails only the fallback test, and clearing it only for this claim fails the three that pin the gossip protection; keying the in-flight claim on the transaction hash fails only the sibling test.

`gofmt`, `go build`, `go vet` clean; `pkg/utils`, `processor`, `indexer`, `pkg/queue/rabbitmq` and the root package all pass under `-race`. `pkg/repo` is excluded — its testcontainers tests need Docker and do not build here, which predates this change.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/relayer/pkg/utils/sending_backend.go [hunk 1]
```diff
@@ -951,6 +1058,48 @@ func (b *SendingBackend) recordSuccess(index int, nonce uint64) {
...
+	for len(b.accepted[index]) > maxTrackedAcceptedNonces {
+		stalest := nonce
+
+		for tracked := range b.accepted[index] {
+			if tracked < stalest {
+				stalest = tracked
+			}
+		}
+
+		delete(b.accepted[index], stalest)
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/relayer/pkg/utils/sending_backend.go]: Fixed in a23827804. The fixed 256-entry eviction is gone: exact endpoint/nonce/claim acceptances are retained regardless of QUEUE_PREFETCH_COUNT and pruned only after DEST_RPC_URL observes a receipt for that nonce or a later one. A late fee-bump success also cannot restore already-mined state. Added regressions for 300 simultaneously live acceptances, receipt-driven pruning, and the receipt/send race.

File: packages/relayer/pkg/utils/sending_backend.go [hunk 2]
```diff
@@ -373,8 +415,27 @@ func (b *SendingBackend) SendTransaction(ctx context.Context, tx *types.Transact
...
+			// transaction to the endpoints behind it leaves the stale variant live in a second
+			// builder's pool — the thing steering a resend to its holder exists to avoid — and
+			// earns a refusal from a relay that did nothing wrong: the receipt lands while a
+			// resubmission is still in flight, and the next endpoint simulates a claim that is
+			// now done. Nothing resets that count while the holder keeps taking every claim
+			// first, so three claims took a healthy backup out of rotation. Returning nil is also
+			// the honest answer for the transaction manager, which is what arms the fee bump and
+			// what TxNotInMempoolTimeout reads: the transaction is at a relay.
+			if answer == heldThisClaim {
+				return nil
```
[the remark this answers is not in the document]
Comment (davidtaikocha) [reply packages/relayer/pkg/utils/sending_backend.go:468]: Documented in a23827804. The README now separates high-water holder ordering from exact endpoint/nonce/claim tracking, states that the exact record has no fixed cap and is pruned by observed receipts, and documents the routing/fallback behavior for same-claim, other-claim, and unattributed held responses.

**Review by smtmfft — CHANGES_REQUESTED** [review]
Blocking on exact head `a23827804d778f990f9bd44063d739c31f17637a`.

1. The receipt/held-response interleaving still reproduces the backup-penalty bug this PR is meant to fix. A private endpoint can accept nonce N, then a fee-bumped resend can already be waiting on that holder while `TransactionReceipt` observes the mined variant. [`clearAcceptedThrough`](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/pkg/utils/sending_backend.go#L531-L557) deletes the exact acceptance before the holder answers; [`recordHeldNonce`](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/pkg/utils/sending_backend.go#L1013-L1022) then sees neither endpoint nor claim identity, classifies the answer as `heldUnattributed`, and the send continues to the backup. The backup can reject the now-DONE claim and be charged/tripped exactly as before. `minedThrough` only prevents a late success from restoring the record; it is not consulted by the held-response path. In an isolated checkout I reproduced this deterministically by blocking the holder's second RPC call, returning the receipt, then releasing an `already known` response: the backup received the transaction and the regression assertion failed. Please preserve enough exact state for an already-in-flight send to still end at its holder, and cover this ordering. The same lifecycle also needs to tolerate a receipt disappearing before confirmation; cleanup is currently irreversible.

2. A public-fallback receipt cannot release the private acceptance records. Successful private sends populate `acceptedTxNonces`, but the successful [all-refused public fallback](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/pkg/utils/sending_backend.go#L471-L491) does not. When a recycled claim lands publicly, [`clearAcceptedThrough`](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/pkg/utils/sending_backend.go#L535-L538) cannot recover its nonce and returns, leaving the earlier private claim in both exact maps. Those maps intentionally have no cap, so repeated public replacements can accumulate stale records for the process lifetime. A second deterministic test drove the existing held-other-claim path through public fallback, returned that public transaction's receipt, and both stale records remained. Please make public-success receipts identify their nonce (or otherwise clean safely) and add this regression case.

3. The duplicate claim is released before its queue disposition completes. [`processMessage` defers `endProcessing`](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/processor/process_message.go#L124-L130), so it runs before the event loop calls [`handleProcessMessageResult`](https://github.com/taikoxyz/taiko-mono/blob/a23827804d778f990f9bd44063d739c31f17637a/packages/relayer/processor/processor.go#L568-L569). With configurable prefetch greater than one, a duplicate can reacquire the hash while Ack/Nack or transient republish is delayed and start processing concurrently, contrary to the new "in-flight delivery is authoritative" invariant. Please hold the claim through the disposition and add a test with a blocked disposition plus a duplicate delivery.

The existing focused packages still pass under `go test -race ./packages/relayer/pkg/utils ./packages/relayer/processor`; these are uncovered interleavings/state-lifetime gaps rather than data-race detector failures.

**Comment by davidtaikocha** [comment]
All three reproduced before being fixed, and each fix is mutation-checked to fail only the test that pins it. Head is now `278702299`.

**1 — receipt and held response interleaving.** Reproduced: after `TransactionReceipt` released the exact record, the holder's `already known` classified as `heldUnattributed` and the send continued to a healthy backup, which received the transaction.

Fixed by keeping the record for the nonce that mined rather than releasing through it: `clearAcceptedThrough` now prunes strictly below. That record is precisely the one an already-in-flight resend needs, and the next nonce to mine releases it, so at most one settled nonce is retained.

That is the second shape this fix took, and the first one is worth recording. I first made a mined nonce end the send on its own — the account's nonces execute in order, so nothing at that nonce can land again. It closed the reproduction, but it answers without knowing which claim the endpoint holds, which is the same watermark mistake this branch already fixed once: a nonce that a reorg made live again and that was then recycled to a different claim would read as "mined, so end the send" and strand the new claim. Keeping the exact record instead answers the question with the claim identity still attached, which also covers the reorg point in your last sentence — the state a resend needs is not thrown away at the receipt, so its disappearance before confirmation has nothing to reverse.

**2 — public-fallback receipts.** Reproduced: a claim landing through the public fallback resolved to no nonce, so `clearAcceptedThrough` returned without releasing anything and the earlier private record stayed in both maps, uncapped.

Every transaction this backend puts on the wire is now indexed by hash, not only the ones a private endpoint accepted, so any landing path is resolvable to its nonce. The field is `sentTxNonces` rather than `acceptedTxNonces` now, because it holds transactions no endpoint accepted and a name that says otherwise is how the next version of this bug gets written.

**3 — claim released before the disposition.** Confirmed in the call sequence: `processMessage` returns, its defer releases, and only then does `handleProcessMessageResult` ack, nack or republish.

The claim now spans the whole delivery. `processDelivery` takes it, runs `processMessage`, settles the result and releases only then; a body that will not decode never held one, so its release is a no-op and the defer is unconditional. The test uses a queue that blocks inside `Ack` and asserts a duplicate arriving in that window is still refused — it fails against the previous shape.

**Test-method note, since it affected the previous round.** My first version of the sibling-delivery test seeded the in-flight map directly and passed under its own mutation: it only proved the test agreed with itself about the key. The claim is now taken through the same derivation production uses, which is why `beginProcessing`/`endProcessing` exist as a seam at all.

`gofmt`, `go build`, `go vet` clean; `pkg/utils`, `processor`, `indexer`, `pkg/queue/rabbitmq` and the root package pass under `-race`.

**Review by smtmfft — CHANGES_REQUESTED** [review]
Re-reviewing exact head `278702299da1081284a53e4bb439290d9e7a0334`. The new commits fix the normal receipt/held-response crossing, index successful public fallback sends, and hold a claim through the normal Ack/Nack/republish path. Three lifecycle gaps remain blocking:

1. The reorg path still leaves the holder identity stale. [`clearAcceptedThrough`](https://github.com/taikoxyz/taiko-mono/blob/278702299da1081284a53e4bb439290d9e7a0334/packages/relayer/pkg/utils/sending_backend.go#L561-L579) retains `accepted[N]` but also advances a monotonic `minedThrough = N` on the first observed receipt. The tx manager calls this wrapper before it checks confirmation depth, and has a `TxNotMined` path when that receipt disappears. After such a reorg, if nonce N is recycled for a different claim B and the private endpoint accepts B, [`recordSuccess`](https://github.com/taikoxyz/taiko-mono/blob/278702299da1081284a53e4bb439290d9e7a0334/packages/relayer/pkg/utils/sending_backend.go#L1147-L1153) returns before replacing claim A's retained exact record because `N <= minedThrough`. B's next held response is therefore classified as `heldOtherClaim` and continues to the backup even though that endpoint demonstrably accepted B. I reproduced this deterministically: accept A(N), observe A's receipt, accept distinct B(N) after the simulated reorg, then return `already known` for B; the backup received B and the regression assertion failed. Please make post-reorg acceptance able to refresh the exact identity without reintroducing the original late-success race, and cover this sequence.

2. `sentTxNonces` still has an unbounded same-nonce lifecycle. Every privately accepted fee variant adds a distinct hash at [`recordSuccess`](https://github.com/taikoxyz/taiko-mono/blob/278702299da1081284a53e4bb439290d9e7a0334/packages/relayer/pkg/utils/sending_backend.go#L1168-L1169), while cleanup removes only entries with nonce strictly below a later mined nonce. If private nonce N never lands, `TX_SEND_TIMEOUT` resets the transaction manager's nonce state and later delivery attempts reuse N; no N+1 receipt can exist to collect those hashes, so they accumulate for the process lifetime. Indexing public sends fixes the specific public-receipt lookup gap, but it does not bound stuck/recycled-nonce variants. Please add a bounded/collectable lifecycle that still preserves every receipt lookup which can remain live.

3. Normal disposition now keeps the claim, but panic disposition still releases it early. If `processMessage` or settlement panics, [`processDelivery`'s deferred release](https://github.com/taikoxyz/taiko-mono/blob/278702299da1081284a53e4bb439290d9e7a0334/packages/relayer/processor/processor.go#L574-L590) runs while unwinding before the outer [`recover` performs its Nack](https://github.com/taikoxyz/taiko-mono/blob/278702299da1081284a53e4bb439290d9e7a0334/packages/relayer/processor/processor.go#L557-L565). A duplicate can reacquire the message while that Nack is blocked or failing. Please keep recovery/disposition inside the claim lifetime and cover a panic plus blocked Nack.

The exact-head suites still pass under `go test -race ./packages/relayer/pkg/utils ./packages/relayer/processor`; the first issue is an uncovered state-machine regression, not a race-detector failure.

**Comment by davidtaikocha** [comment]
All three reproduced before being fixed. Head is now `9c808321d`.

**1 — reorged nonce keeps a stale holder identity.** Correct, and my previous reply overstated: I wrote that retaining the mined nonce's record "covers the reorg point", which is only true while the nonce keeps the same claim. Once it is recycled, the retained record names the claim the reorg removed and the endpoint that demonstrably accepted the new one is read as holding someone else's. Reproduced exactly as described — the backup received B.

`recordSuccess` now compares the claim instead of the nonce alone. A success carrying the claim already recorded at that nonce is the stale fee variant the original guard was for, and is still skipped; anything else means the chain moved and replaces the record. The late-success regression from the earlier round still passes unchanged.

**2 — unbounded same-nonce lifecycle in the index.** Two separate causes, so two fixes.

A nonce executes once, so when a receipt arrives every *other* hash recorded at that nonce is dead forever. Those are now collected with it, keeping only the hash that landed, which the transaction manager reads more than once. That removes the fee-variant fan-out at every nonce that mines.

The stuck case has no receipt to do that — nothing above the nonce can mine either — so it takes a cap, at 1024. I want to be explicit that this does not meet "preserves every receipt lookup which can remain live" in the strict sense: at that point every entry shares the stuck nonce and any of them could in principle be the one a relay still holds. What makes it safe is that a lost lookup is not a correctness failure. `clearAcceptedThrough` returns without releasing anything that time, and the next receipt that does resolve releases everything below it. Eviction takes the lowest nonce, which under a stuck nonce leaves the oldest variants — the low-fee ones its own resends have already replaced. If you would rather have an exact structure than a bound with that argument behind it, say so and I will build it.

**3 — panic disposition outside the claim.** Confirmed in the ordering: `processDelivery`'s deferred release runs while unwinding, the outer recover's nack runs after. The recovery moves inside `processDelivery`, ahead of the release.

The first version of that test was not good enough to merit the fix. It let the panic escape, so the mutation crashed the test binary instead of failing an assertion — that shows the recovery moved, not that the claim outlived the settlement. The test now stands in for the old outer recovery in its own goroutine, so the two shapes differ only in where the claim ends, and reverting the fix fails the assertion cleanly.

Each fix is mutation-checked and each mutation fails only the test that pins it: nonce-only skip → the reorg test; no same-nonce collection → the dead-variant test; no cap → the bound test; recovery back outside the claim → the panic test.

`gofmt`, `go build`, `go vet` clean; `pkg/utils`, `processor`, `indexer`, `pkg/queue/rabbitmq` and the root package pass under `-race`.

**Review by smtmfft — CHANGES_REQUESTED** [review]
Re-reviewing the reply and exact head `7f5723c89fe0a18cd56b63056d21c853f57f6203`. The direct post-reorg A→B transition and the panic-Nack claim lifetime are now fixed, and their new regressions pass. Two ordering/lifecycle gaps remain blocking:

1. Comparing only the current claim does not order overlapping pre- and post-reorg successes. Consider a fee-bumped private RPC for old claim A(N) that is already in flight: A's receipt is observed, the receipt is reorged out, and the endpoint then accepts recycled claim B(N), so B correctly replaces A in `accepted[index][N]`. If the old A RPC response returns success after that, [`recordSuccess`](https://github.com/taikoxyz/taiko-mono/blob/7f5723c89fe0a18cd56b63056d21c853f57f6203/packages/relayer/pkg/utils/sending_backend.go#L1191-L1206) sees `accepted[index][N] != claim(A)` and treats the stale A result as chain movement, overwriting B at lines 1222–1225. B's next `already known` is then classified as `heldOtherClaim` and goes to the backup even though this endpoint accepted B. I reproduced this deterministically with A receipt → B acceptance → late fee-bumped A success → B held response; the backup received B and the assertion failed. The new reorg test stops before the late A response. Please add an ordering/epoch that prevents an older request result from overwriting a newer acceptance, and cover this exact sequence.

2. The new cap can evict the receipt lookup needed to drain the independently uncapped exact records. [`boundSentNonces`](https://github.com/taikoxyz/taiko-mono/blob/7f5723c89fe0a18cd56b63056d21c853f57f6203/packages/relayer/pkg/utils/sending_backend.go#L572-L586) chooses arbitrarily among equal lowest-nonce variants because Go map iteration carries no age order; it therefore does not implement the comment's claim that old low-fee variants are what remain/leave. Any accepted variant can in principle be the one that lands. If its hash was evicted, [`clearAcceptedThrough`](https://github.com/taikoxyz/taiko-mono/blob/7f5723c89fe0a18cd56b63056d21c853f57f6203/packages/relayer/pkg/utils/sending_backend.go#L592-L599) returns without releasing older `accepted` entries. I reproduced this by taking an evicted variant at N as the receipt hash: the exact record at N-1 remained. The current cap test proves only `len(sentTxNonces) <= 1024`; it does not prove cleanup progress or bound the `accepted` maps after lookup misses. Please make eviction follow a justified deterministic invariant, or make receipt cleanup independent of a cache hit, and add the missing receipt-after-eviction regression.

The latest exact-head suites still pass under `go test -race ./packages/relayer/pkg/utils ./packages/relayer/processor`. DeepSeek's repeated normal-refusal note is intentional pre-existing ceiling behavior and is not part of this block.

**Comment by davidtaikocha** [comment]
Both reproduced before being fixed. Head is now `7acf6274b`.

**1 — ordering between overlapping pre- and post-reorg successes.** Correct: comparing claims answers *what* an endpoint took and nothing about *when*, so a late result and a genuine reorg are indistinguishable. Reproduced through the public API by stalling the fee-bumped A send inside the endpoint, accepting B while it hung, then releasing it — the backup received B.

Each acceptance now carries the send that learned it. `SendTransaction` takes a monotonic ticket before it offers the transaction anywhere, so two sends of one nonce are ordered by when they started rather than by when their answers came back, and `recordSuccess` ignores a result older than the record already in place. The two earlier guards are unchanged and still needed: a repeat of the recorded claim under a mined nonce is still the stale fee variant, and a different claim from a *newer* send is still the chain having moved.

**2 — eviction and cache-dependent cleanup.** Both halves were right, and the second is the one that mattered.

Eviction is now deterministic: the oldest send goes first. Every entry under a stuck nonce shares that nonce, so the nonce cannot order them and map iteration carries no age — the send number is the only thing that does, and it identifies the resend a relay is least likely still to hold. The comment now describes what the code does.

That still cannot promise the evicted hash was not the one that lands, so I took the other option as well rather than relying on the argument: a receipt this index cannot resolve is settled against the account's nonce on `DEST_RPC_URL`. Releasing no longer depends on the cache hitting at all, which is what the deliberately uncapped exact records need. `PendingNonceAt` is now overridden to remember the account, since that was the only place this backend could learn which address it sends for.

The regression you asked for is there: an entry is deleted from the index, its receipt read, and the older exact record must still be released while the settled nonce's own record is kept.

**On the cap test.** You were right that it proved only `len <= 1024`. I confirmed that directly — inverting the eviction to drop the *newest* send passed the whole suite before I strengthened it. It now pins which entries survive, and that mutation fails.

Mutation checks, each failing only its own test: dropping the send-ordering guard → the late-success test; removing the chain fallback → the evicted-lookup test; evicting newest instead of oldest → the cap test.

`gofmt`, `go build`, `go vet` and `golangci-lint` with the repo config are clean; `pkg/utils`, `processor`, `indexer`, `pkg/queue/rabbitmq` and the root package pass under `-race`.

Unrelated to this branch: `test-relayer` has been failing on `pkg/repo`'s testcontainers, which times out pulling `mysql:latest` inside the 5-minute context in `containers_test.go`. It passed on `9c808321d`, `pkg/repo` is untouched here, and three re-runs reproduce it. Happy to pin that image and separate the pull from the readiness budget in its own PR if you want it.

**Review by smtmfft — CHANGES_REQUESTED** [review]
Re-reviewing exact head `7acf6274bbb546160cec8becfb65160f9dac07fa`. The send-ticket ordering fixes the previous late pre-reorg success overwrite, but two receipt-lifecycle gaps remain blocking.

1. [`SendingBackend.PendingNonceAt`](https://github.com/taikoxyz/taiko-mono/blob/7acf6274bbb546160cec8becfb65160f9dac07fa/packages/relayer/pkg/utils/sending_backend.go#L630-L641) is the only method that records `account` / `hasAccount`. The pinned txmgr does not use it when allocating a transaction nonce: [`signWithNextNonce` calls `backend.NonceAt(ctx, cfg.From, nil)`](https://github.com/taikoxyz/optimism/blob/5490c5186828/op-service/txmgr/txmgr.go#L510-L519). `SendingBackend` does not override `NonceAt`, so that call goes directly through the embedded backend and leaves `hasAccount == false`. When an evicted transaction later produces a receipt, [`clearAcceptedBelowChain` therefore returns immediately](https://github.com/taikoxyz/taiko-mono/blob/7acf6274bbb546160cec8becfb65160f9dac07fa/packages/relayer/pkg/utils/sending_backend.go#L655-L662), and the older uncapped exact acceptance records remain.

The new regression masks the integration gap by [explicitly calling `b.PendingNonceAt`](https://github.com/taikoxyz/taiko-mono/blob/7acf6274bbb546160cec8becfb65160f9dac07fa/packages/relayer/pkg/utils/sending_backend_test.go#L2419-L2421), which production does not do. I reproduced the production shape deterministically: accept nonces 40 and 41, call `b.NonceAt(..., nil)` as this txmgr does, evict nonce 41's hash, then return its receipt. The record for nonce 40 remains and the assertion fails.

2. A later nonce's receipt can still erase the exact holder state needed by an earlier resend that is already in flight. Concurrent txmgr sends may have accepted N and N+1, while a fee-bumped resend of N is waiting on its holder. Observing the N+1 receipt is valid because account nonces execute in order, but [`clearAcceptedThrough` deletes every acceptance below N+1](https://github.com/taikoxyz/taiko-mono/blob/7acf6274bbb546160cec8becfb65160f9dac07fa/packages/relayer/pkg/utils/sending_backend.go#L723-L729). When the holder's delayed `already known` answer for N returns, `recordHeldNonce` sees no exact record, classifies it as unattributed, and [continues to the backup](https://github.com/taikoxyz/taiko-mono/blob/7acf6274bbb546160cec8becfb65160f9dac07fa/packages/relayer/pkg/utils/sending_backend.go#L467-L474). I reproduced this deterministically with accepted N/N+1, a stalled N resend, the N+1 receipt, then the holder response: the backup received N and the assertion failed. The existing test proves that the older record is deleted but does not put an earlier resend in flight.

Please learn the sender from the actual `NonceAt` path (preserving its block-number semantics), or pass the configured sender into the wrapper explicitly, and make the cache-miss test exercise the same call used by the pinned txmgr. The exact acceptance lifecycle also needs to retain or otherwise recognize every resend that was already offered to its holder when a later receipt performs cleanup. The exact-head race suites otherwise pass, and the previous late-success ordering regression now passes.
